Readme.md
# Kette
Concatenative Stack-Based Programming Language

## Usage
1. compile the compiler: `fasm kette.asm`
2. compile file to asm: `./kette <file> <output-fasm>`
3. compile asm: `fasm <output-fasm> <output>`
4. run code: `./<output>`


## Features & TODO list
** everything marked with prio 0 and 1 is "required" before selfhosting
- [x] if else
- [x] loops
- [x] procs
- [x] constants (constants are special case of procs in this lang lmao)
- [x] strings | constants work already | PRIO 0
- [ ] string formatting (\n & number / string insert & unicode) | PRIO 1
- [x] comments | PRIO 0
- [ ] single error handling | PRIO 1 | multi error handling will be added in self host
- [x] command line arguments
- [ ] command line arguments code | code in place, needs accessibility | PRIO 1
- [ ] ptr | PRIO 0
- [ ] mutable globals | PRIO 1
- [ ] syntax highlighting VSC + VIM | PRIO 2
- [ ] types (int, string, ptr)
- [ ] static typechecking
- [ ] arrays
- [x] syscalls
- [ ] match (on types & values) 
- [ ] modules | PRIO 1
- [ ] simple datastructures ex. HashMap, Dynamic Array
- [ ] self hosted | PRIO 2
- [ ] some optimizations
- [ ] proc options e.g. "inline", "c"
- [ ] memory management
- [ ] c-interop
- [ ] some OOP
- [ ] Stack Effects & Pure Functions
- [ ] multithreading
- [ ] REPL
- [ ] AArch64 Mac & Windows and x64 Windows backend
- [ ] javascript backend lol
- [ ] package manager

## Compiler Design
### 1. Lexing
1. loop through input string
2. return pointer to substring with length, skipping whitespaces
3. while looping, increment line and col counter, reset col counter when newline is reached
4. strings are parsed from " to ", keeping all newlines and spaces
5. substrings are matched in order by string, keyword, number, "word" and converted to the token format

#### Notes
- Too large numbers and non closing " can be detected here
- TODO: currently numbers are expected at the end -> change that and do symbols instead

### 2. Token Format
a single token is 48 bytes long and structured the following:
1. [1b] token kind  | 0
2. [4b] file        | 1
3. [4b] line        | 5
4. [4b] column      | 9
5. [27b] data       | 13

#### Notes
- index 9 may be used for subvariance that may be detected for some tokens later
- some spaces are reused e.g. a do writes last while in it first which is then being changed to end later
- most tokens are 9 bytes and 17 bytes long but for easier iteration they are all stretched to 32 bytes

### Primitive-Cross-Referencing
- references if with else and end
- references while with do and end
- references proc with end

#### Notes
- detects missing end, if and while

### Symbol Cross-Referencing
- references global variabls
- references procedure calls

#### Notes
- detects missing symbols

### Assembly Generation
- generates the assembly for x64 linux
- TODO: add assembly for Windows (probably only syscalls, or include Win32.h?)
- TODO: Add AArch64 support


## Assembly Notes

### SysV ABI
#### Parameters:
1. rdi
2. rsi
3. rdx
4. rcx
5. r8
6. r9
7. stack

Return:
1. rax
2. rdx
3. stack | preallocated in caller stack frame and passed as pointer

#### Preserve:
- rbx
- rsp | unusable, only for stack frames 
- rbp | unusable, only for stack frames
- rip | do not touch
- r12
- r13
- r14
- r15

#### Whatever:
(Parameters and RCX)
- rax
- rdi
- rsi
- rdx
- rcx  
- r8
- r9
- r10
- r11

### Linux
#### Notes
syscalls may override:
- rcx | rip of syscall
- r11 | RFLAGS

#### Parameter Order
1. rdi
2. rsi
3. rdx
4. r10
5. r8
6. r9
7. rax | which sycall


kette-language-support/CHANGELOG.md
# Change Log

All notable changes to the "kette-language-support" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

test.asm
format ELF64 executable 3

segment readable executable
entry $
 	mov rdi, Sample
    mov rsi, Size
    call print_string

    mov rdi, 0
    mov rax, 60
    syscall


; input:
;   rdi: ptr string
;   rsi: len string
print_string:
    push    rdi
    push    rax
    push    rdx
    push    rcx
    push    rsi

    mov     rdx, rsi
    mov     rsi, rdi
    mov     rdi, 1
    mov     rax, 1
    syscall
    
    pop     rsi
    pop     rcx
    pop     rdx
    pop     rax
    pop     rdi
    ret

; prints uint
; input:
;   rdi: value
; output:
;   rdi: value (unchanged)
print_uint:
    push    rbp
    mov     rbp, rsp

    push    rdi

    lea     rsi, [rsp] ; save "base" addr for uitds
    sub     rsp, 20 ; max digits + new line

    call    uitds
    
    mov     r8, 20
    sub     r8, rdx
    add     rsp, r8

    mov     rsi, rax
    mov     rdx, rdx ; rdx is already len
    mov     rdi, 1
    mov     rax, 1
    syscall

    pop     rdi
    mov     rsp, rbp
    pop     rbp
    ret

; input:
;   rdi: value
;   rsi: ptr string (only alloc)
; output:
;   rax: ptr string (modified)
;   rdx: uint = len
uitds:
    push    rdi ; save value for second loop
    mov     rax, rdi ; use for div
    mov     r8 , 10 ; use as divisor
    
    xor     rcx, rcx ; be sure rcx is 0
    .loop_uitds_count:
        xor     rdx, rdx ; reset rdx for div
        div     r8 ; rax / 10 = rax, rest: rdx
        inc     rcx
        test    rax, rax
        jnz     .loop_uitds_count

    pop     rax ; pop saved value for second loop
    sub     rsi, rcx ; "allocate memory" within stack
    push    rsi ; store pointer to string start
    push    rcx ; store string len

    .loop_uitds_write:
        xor     rdx, rdx
        div     r8
        add     rdx, 48 ; rest div = digit, num + 48 => ascii num
        ; move only a byte into the stack
        ; -1 because rcx is len not index
        mov     BYTE [rsi + rcx - 1], dl 
        dec     rcx
        jnz     .loop_uitds_write
    
    pop     rdx ; take string len
    pop     rax ; take string pointer
    ret


segment readable writable
Sample db "Sample text of mixed alphabets:",10
       DB "Éireannach (Eireannach in western European alphabet)",10
       DB "Čapek (Capek in central European alphabet)",10
       DB "Ørsted (Oersted in Nordic alphabet)",10
       DB "Aukštaičių (Aukshtaiciu in Baltic alphabet)",10
       DB "Ὅμηρος (Homer in Greek alphabet)",10
       DB "Yumuşak ğ (Yumushak g in Turkish aplhabet)",10
       DB "Maðkur (Mathkur in Icelandic alphabet)",10
       DB "דגבא (ABGD in Hebrew alphabet)",10
       DB "Достоевский (Dostoevsky in Cyrillic alphabet)",10, 0x00, 0xF1

Size = $ - Sample

arith.gen.asm
format ELF64 executable 3

segment readable executable

dump_uint:
mov     r8, -3689348814741910323
sub     rsp, 40
mov     BYTE [rsp+19], 10
lea     rcx, [rsp+18]
.L2:
mov     rax, rdi
mul     r8
mov     rax, rdi
shr     rdx, 3
lea     rsi, [rdx+rdx*4]
add     rsi, rsi
sub     rax, rsi
add     eax, 48
mov     BYTE [rcx], al
mov     rax, rdi
mov     rdi, rdx
mov     rdx, rcx
sub     rcx, 1
cmp     rax, 9
ja      .L2
lea     rax, [rsp+20]
mov     edi, 1
mov     rcx, rax
sub     rcx, rdx
sub     rdx, rax
lea     rsi, [rsp+20+rdx]
mov     rdx, rcx
mov     rax, 1
syscall
add     rsp, 40
ret

entry $
; -- STARTUP -- 
mov rax, RETURN_STACK_END
mov [RETURN_STACK_PTR], rax

; -- PROC DECLERATION --
; - skip -
jmp _Addr0
_Proc0:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 1
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr0:
; -- PROC DECLERATION --
; - skip -
jmp _Addr1
_Proc1:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 1
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr1:
; -- PROC DECLERATION --
; - skip -
jmp _Addr2
_Proc2:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH STRING --
push CONST_STRING_1_LEN
push CONST_STRING_1
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr2:
; -- PROC DECLERATION --
; - skip -
jmp _Addr3
_Proc3:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc1
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc0
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- SYSCALL3 --
pop rax
pop rdi
pop rsi
pop rdx
syscall
push rax
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr3:
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc2
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc3
mov [RETURN_STACK_PTR], rsp
mov rsp, rax

mov rdi, 0
mov rax, 60
syscall

; -- CONST DATA --
segment readable

CONST_STRING_1 db "test.ket"
CONST_STRING_1_LEN = $ - CONST_STRING_1
; -- MUTABLE DATA --
segment readable writable

; -- RETURN STACK --
RETURN_STACK_PTR rq 1
RETURN_STACK rq 512
RETURN_STACK_END:


examples/fizzbuzz.asm
format ELF64 executable 3

segment readable executable

dump_uint:
mov     r8, -3689348814741910323
sub     rsp, 40
mov     BYTE [rsp+19], 10
lea     rcx, [rsp+18]
.L2:
mov     rax, rdi
mul     r8
mov     rax, rdi
shr     rdx, 3
lea     rsi, [rdx+rdx*4]
add     rsi, rsi
sub     rax, rsi
add     eax, 48
mov     BYTE [rcx], al
mov     rax, rdi
mov     rdi, rdx
mov     rdx, rcx
sub     rcx, 1
cmp     rax, 9
ja      .L2
lea     rax, [rsp+20]
mov     edi, 1
mov     rcx, rax
sub     rcx, rdx
sub     rdx, rax
lea     rsi, [rsp+20+rdx]
mov     rdx, rcx
mov     rax, 1
syscall
add     rsp, 40
ret

entry $
; -- STARTUP -- 
mov rax, RETURN_STACK_END
mov [RETURN_STACK_PTR], rax

mov rax, rsp
mov [ARGS_COUNT], rax
add rax, 8
mov [ARGS_PTR], rax

; -- PROC DECLERATION --
; - skip -
jmp _Addr0
_Proc0:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 1
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr0:
; -- PROC DECLERATION --
; - skip -
jmp _Addr1
_Proc1:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 1
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr1:
; -- PROC DECLERATION --
; - skip -
jmp _Addr2
_Proc2:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc0
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- SYSCALL3 --
pop rax
pop rdi
pop rsi
pop rdx
syscall
push rax
; -- DROP --
pop r15
xor r15, r15
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr2:
; -- PROC DECLERATION --
; - skip -
jmp _Addr3
_Proc3:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc1
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc2
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr3:
; -- PUSH --
push 15
; -- WHILE --
_Addr8:
; -- DUP --
pop rax
push rax
push rax
; -- PUSH --
push 0
; -- GREATER --
pop rbx
pop rax
xor rcx, rcx
cmp rax, rbx
mov rdx, 1
cmovg rcx, rdx
push rcx
; -- DO --
pop rax
cmp rax, 0
jz _Addr9
; -- IF --
; -- DUP --
pop rax
push rax
push rax
; -- PUSH --
push 3
; -- MOD --
pop rbx
pop rax
xor rdx, rdx
div rbx
push rdx
; -- PUSH --
push 0
; -- EQUAL --
pop rbx
pop rax
xor rcx, rcx
cmp rax, rbx
mov rdx, 1
cmove rcx, rdx
push rcx
; -- OVER --
pop rax
pop rbx
push rbx
push rax
push rbx
; -- PUSH --
push 5
; -- MOD --
pop rbx
pop rax
xor rdx, rdx
div rbx
push rdx
; -- PUSH --
push 0
; -- EQUAL --
pop rbx
pop rax
xor rcx, rcx
cmp rax, rbx
mov rdx, 1
cmove rcx, rdx
push rcx
; -- AND --
pop rbx
pop rax
mov rcx, 1
mov rdx, 0
and rax, rbx
cmp rax, 0
cmovnz rdx, rcx
push rdx
; -- THEN -- 
pop rax
cmp rax, 0
jz _Addr4
; -- PUSH STRING --
push CONST_STRING_1_LEN
push CONST_STRING_1
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc3
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- PUSH --
push 0
; -- DUMP --
pop rdi
call dump_uint
; -- ELSE --
jmp _Addr7
_Addr4:
; -- IF --
; -- DUP --
pop rax
push rax
push rax
; -- PUSH --
push 3
; -- MOD --
pop rbx
pop rax
xor rdx, rdx
div rbx
push rdx
; -- PUSH --
push 0
; -- EQUAL --
pop rbx
pop rax
xor rcx, rcx
cmp rax, rbx
mov rdx, 1
cmove rcx, rdx
push rcx
; -- THEN -- 
pop rax
cmp rax, 0
jz _Addr5
; -- PUSH STRING --
push CONST_STRING_2_LEN
push CONST_STRING_2
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc3
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- PUSH --
push 0
; -- DUMP --
pop rdi
call dump_uint
; -- END --
_Addr5:
; -- IF --
; -- DUP --
pop rax
push rax
push rax
; -- PUSH --
push 5
; -- MOD --
pop rbx
pop rax
xor rdx, rdx
div rbx
push rdx
; -- PUSH --
push 0
; -- EQUAL --
pop rbx
pop rax
xor rcx, rcx
cmp rax, rbx
mov rdx, 1
cmove rcx, rdx
push rcx
; -- THEN -- 
pop rax
cmp rax, 0
jz _Addr6
; -- PUSH STRING --
push CONST_STRING_3_LEN
push CONST_STRING_3
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc3
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- PUSH --
push 0
; -- DUMP --
pop rdi
call dump_uint
; -- END --
_Addr6:
; -- END --
_Addr7:
; -- DUP --
pop rax
push rax
push rax
; -- DUMP --
pop rdi
call dump_uint
; -- PUSH --
push 1
; -- SUB --
pop rbx
pop rax
sub rax, rbx
push rax
; -- END --
jmp _Addr8
_Addr9:

mov rdi, 0
mov rax, 60
syscall

; -- CONST DATA --
segment readable

CONST_STRING_1 db "FizzBuzz", 0
CONST_STRING_1_LEN = $ - CONST_STRING_1 - 1
CONST_STRING_2 db "Fizz", 0
CONST_STRING_2_LEN = $ - CONST_STRING_2 - 1
CONST_STRING_3 db "Buzz", 0
CONST_STRING_3_LEN = $ - CONST_STRING_3 - 1


; -- MUTABLE DATA --
segment readable writable

; -- RETURN STACK --
RETURN_STACK_PTR rq 1
RETURN_STACK rq 1028
RETURN_STACK_END:
ARGS_COUNT rq 1
ARGS_PTR rq 1


examples/arithmetic.asm
format ELF64 executable 3

segment readable executable

dump_uint:
mov     r8, -3689348814741910323
sub     rsp, 40
mov     BYTE [rsp+19], 10
lea     rcx, [rsp+18]
.L2:
mov     rax, rdi
mul     r8
mov     rax, rdi
shr     rdx, 3
lea     rsi, [rdx+rdx*4]
add     rsi, rsi
sub     rax, rsi
add     eax, 48
mov     BYTE [rcx], al
mov     rax, rdi
mov     rdi, rdx
mov     rdx, rcx
sub     rcx, 1
cmp     rax, 9
ja      .L2
lea     rax, [rsp+20]
mov     edi, 1
mov     rcx, rax
sub     rcx, rdx
sub     rdx, rax
lea     rsi, [rsp+20+rdx]
mov     rdx, rcx
mov     rax, 1
syscall
add     rsp, 40
ret

entry $
; -- STARTUP -- 
mov rax, RETURN_STACK_END
mov [RETURN_STACK_PTR], rax

mov rax, rsp
mov [ARGS_COUNT], rax
add rax, 8
mov [ARGS_PTR], rax

; -- PUSH --
push 5
; -- PUSH --
push 7
; -- ADD --
pop rbx
pop rax
add rax, rbx
push rax
; -- PUSH --
push 3
; -- ADD --
pop rbx
pop rax
add rax, rbx
push rax
; -- DUMP --
pop rdi
call dump_uint
; -- PUSH --
push 5
; -- PUSH --
push 7
; -- PUSH --
push 3
; -- ADD --
pop rbx
pop rax
add rax, rbx
push rax
; -- ADD --
pop rbx
pop rax
add rax, rbx
push rax
; -- DUMP --
pop rdi
call dump_uint
; -- PUSH --
push 19
; -- PUSH --
push 4
; -- MODIV --
pop rbx
pop rax
xor rdx, rdx
div rbx
push rdx
push rax
; -- DUMP --
pop rdi
call dump_uint
; -- DUMP --
pop rdi
call dump_uint
; -- PUSH --
push 3
; -- PUSH --
push 1
; -- SUB --
pop rbx
pop rax
sub rax, rbx
push rax
; -- MUL --
pop rbx
pop rax
mul rbx
push rax
; -- PUSH --
push 100
; -- DUMP --
pop rdi
call dump_uint

mov rdi, 0
mov rax, 60
syscall

; -- CONST DATA --
segment readable



; -- MUTABLE DATA --
segment readable writable

; -- RETURN STACK --
RETURN_STACK_PTR rq 1
RETURN_STACK rq 1028
RETURN_STACK_END:
ARGS_COUNT rq 1
ARGS_PTR rq 1


examples/files.asm
format ELF64 executable 3

segment readable executable

dump_uint:
mov     r8, -3689348814741910323
sub     rsp, 40
mov     BYTE [rsp+19], 10
lea     rcx, [rsp+18]
.L2:
mov     rax, rdi
mul     r8
mov     rax, rdi
shr     rdx, 3
lea     rsi, [rdx+rdx*4]
add     rsi, rsi
sub     rax, rsi
add     eax, 48
mov     BYTE [rcx], al
mov     rax, rdi
mov     rdi, rdx
mov     rdx, rcx
sub     rcx, 1
cmp     rax, 9
ja      .L2
lea     rax, [rsp+20]
mov     edi, 1
mov     rcx, rax
sub     rcx, rdx
sub     rdx, rax
lea     rsi, [rsp+20+rdx]
mov     rdx, rcx
mov     rax, 1
syscall
add     rsp, 40
ret

entry $
; -- STARTUP -- 
mov rax, RETURN_STACK_END
mov [RETURN_STACK_PTR], rax

mov rax, rsp
mov [ARGS_COUNT], rax
add rax, 8
mov [ARGS_PTR], rax

; -- PROC DECLERATION --
; - skip -
jmp _Addr0
_Proc0:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 0
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr0:
; -- PROC DECLERATION --
; - skip -
jmp _Addr1
_Proc1:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 1
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr1:
; -- PROC DECLERATION --
; - skip -
jmp _Addr2
_Proc2:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 2
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr2:
; -- PROC DECLERATION --
; - skip -
jmp _Addr3
_Proc3:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 3
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr3:
; -- PROC DECLERATION --
; - skip -
jmp _Addr4
_Proc4:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 8
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr4:
; -- PROC DECLERATION --
; - skip -
jmp _Addr5
_Proc5:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 9
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr5:
; -- PROC DECLERATION --
; - skip -
jmp _Addr6
_Proc6:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 0
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr6:
; -- PROC DECLERATION --
; - skip -
jmp _Addr7
_Proc7:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 1
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr7:
; -- PROC DECLERATION --
; - skip -
jmp _Addr8
_Proc8:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 64
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr8:
; -- PROC DECLERATION --
; - skip -
jmp _Addr9
_Proc9:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 512
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr9:
; -- PROC DECLERATION --
; - skip -
jmp _Addr10
_Proc10:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 0
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr10:
; -- PROC DECLERATION --
; - skip -
jmp _Addr11
_Proc11:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 1
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr11:
; -- PROC DECLERATION --
; - skip -
jmp _Addr12
_Proc12:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 2
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr12:
; -- PROC DECLERATION --
; - skip -
jmp _Addr13
_Proc13:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 0
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr13:
; -- PROC DECLERATION --
; - skip -
jmp _Addr14
_Proc14:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 1
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr14:
; -- PROC DECLERATION --
; - skip -
jmp _Addr15
_Proc15:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 2
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr15:
; -- PROC DECLERATION --
; - skip -
jmp _Addr16
_Proc16:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc1
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- SYSCALL3 --
pop rax
pop rdi
pop rsi
pop rdx
syscall
push rax
; -- DROP --
pop r15
xor r15, r15
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr16:
; -- PROC DECLERATION --
; - skip -
jmp _Addr17
_Proc17:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc14
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc16
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr17:
; -- PROC DECLERATION --
; - skip -
jmp _Addr18
_Proc18:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc2
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- SYSCALL3 --
pop rax
pop rdi
pop rsi
pop rdx
syscall
push rax
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr18:
; -- PROC DECLERATION --
; - skip -
jmp _Addr19
_Proc19:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc12
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- PUSH --
push 0
; -- ROT --
pop rax
pop rbx
pop rcx
push rbx
push rax
push rcx
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc4
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- SYSCALL3 --
pop rax
pop rdi
pop rsi
pop rdx
syscall
push rax
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr19:
; -- PROC DECLERATION --
; - skip -
jmp _Addr20
_Proc20:
; - prepare -
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; - body -
; -- PUSH --
push 0
; -- 2SWAP --
pop rax
pop rbx
pop rcx
pop rdx
push rbx
push rax
push rdx
push rcx
; -- SWAP --
pop rax
pop rbx
push rax
push rbx
; -- DROP --
pop r15
xor r15, r15
; -- ROT --
pop rax
pop rbx
pop rcx
push rbx
push rax
push rcx
; -- SWAP --
pop rax
pop rbx
push rax
push rbx
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc18
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- PROC END --
; - return - 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
ret
; - skip -
_Addr20:
; -- PUSH STRING --
push CONST_STRING_1_LEN
push CONST_STRING_1
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc6
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc20
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- DUP --
pop rax
push rax
push rax
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc19
mov [RETURN_STACK_PTR], rsp
mov rsp, rax
; -- SWAP --
pop rax
pop rbx
push rax
push rbx
; -- CALL PROC -- 
mov rax, rsp
mov rsp, [RETURN_STACK_PTR]
call _Proc17
mov [RETURN_STACK_PTR], rsp
mov rsp, rax

mov rdi, 0
mov rax, 60
syscall

; -- CONST DATA --
segment readable

CONST_STRING_1 db "test.txt", 0
CONST_STRING_1_LEN = $ - CONST_STRING_1 - 1


; -- MUTABLE DATA --
segment readable writable

; -- RETURN STACK --
RETURN_STACK_PTR rq 1
RETURN_STACK rq 1028
RETURN_STACK_END:
ARGS_COUNT rq 1
ARGS_PTR rq 1
